extensions/common/context.cc

```diff
@@ -123,28 +88,37 @@ void getDestinationService(const std::string& dest_namespace,
     return;
   }
 
-  std::vector<std::string_view> parts = absl::StrSplit(cluster_name, '|');
-  if (parts.size() == 4) {
-    *dest_svc_host = std::string(parts[3].data(), parts[3].size());
+  // Get destination service name and host from cluster labels.
+  if (getValue({"cluster_metadata", "filter_metadata", "istio", "services", "0",
+                "name"},
+               dest_svc_name)) {
+    getValue({"cluster_metadata", "filter_metadata", "istio", "services", "0",
+              "host"},
+             dest_svc_host);
+  } else if (use_host_header) {
+    // if cluster metadata cannot be found, fallback to host header.
```

Comment:
I updated the comment here as well as the comment of this function. I think now this will only happen when no cluster is matched for the request. wildcard cluster is handled above.